src/puppetlabs/puppetdb/cli/benchmark.clj

```diff
@@ -503,57 +591,48 @@
 
+    ;; We don't have to worry about these core.async threads during
+    ;; shutdown because they appear to be non-daemon, and so just dead.
     (dotimes [_ threads]
       (start-command-sender base-url command-send-ch rate-monitor-ch))
 
-    (when-not catalogs (log/info (trs "No catalogs specified; skipping catalog submission")))
-    (when-not reports (log/info (trs "No reports specified; skipping report submission")))
-    (when-not facts (log/info (trs "No facts specified; skipping fact submission")))
-
     (if nummsgs
       (do
-        (submit-n-messages hosts nummsgs rand-perc command-send-ch)
+        (submit-messages (random-hosts numhosts pdb-host catalogs reports facts)
+                         nummsgs rand-perc command-send-ch)
         (dotimes [_ threads]
           (>!! command-send-ch :stop))
-        (close! command-send-ch))
-
-      (let [close-to-stop-ch (chan)
-            mq (chan (amq-broker-buffer "benchmark-mq" "benchmark-endpoint"))]
-
-        ;; be sure to clear the temp queue on exit
-        (.addShutdownHook (Runtime/getRuntime) (Thread. #(close! mq)))
-
-        ;; initial queue population
-        (future
-          (println "Populating the queue in the background")
-          (dotimes [n numhosts]
-            (let [h (-> (make-host n)
-                        (assoc :lastrun (rand-lastrun run-interval)))]
-              (>!! mq h)))
-
-          (println "... finished filling the queue, continuing with the simulation."))
-
-        (let [run-interval-minutes (time/in-minutes run-interval)
-              hosts-per-second (/ numhosts (* run-interval-minutes 60))
-              rate-limited-mq-out (chan)]
-          (relay-messages-at-rate mq rate-limited-mq-out hosts-per-second)
-          (pipeline simulation-threads
-                    mq
-                    (map #(update-host % rand-perc (time/now) command-send-ch))
-                    rate-limited-mq-out)
-          (go
-            (<! close-to-stop-ch)
-            (mapv close! [rate-limited-mq-out mq rate-monitor-ch command-send-ch])))
-        close-to-stop-ch))))
-
-(defn chan? [x]
-  (satisfies? clojure.core.async.impl.protocols/Channel x))
-
+        (close! command-send-ch)
+        nil)
+      ;; Run as a daemon thread so it's not just dead during shutdown
+      (let [result (promise)]
+        (doto (Thread.
+               #(try
+                  (deliver result
+                           [(cycle-commands numhosts run-interval rand-perc
+                                            simulation-threads
+                                            rate-monitor-ch command-send-ch
+                                            pdb-host catalogs reports facts)])
+                  (catch Exception ex
+                    (deliver result ex))))
+          (.setDaemon true)
+          .start)
+        (let [x @result]
+          (if (vector? x)
+            (first x)
+            (throw x)))))))
```

Comment:
...and it's really just a redirection of the daemon-thread's result (exception or not) back onto the current thread, since the only reason for the temporary switch to a daemon thread is so that we won't be unceremoniously plonked by shutdown.